### PR TITLE
[SPARK-31256][SQL] DataFrameNaFunctions.drop should work for nested columns

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

#26700 removed the ability to drop a row whose nested column value is null.

For example, for the following `df`:
```
val schema = new StructType()
  .add("c1", new StructType()
    .add("c1-1", StringType)
    .add("c1-2", StringType))
val data = Seq(Row(Row(null, "a2")), Row(Row("b1", "b2")), Row(null))
val df = spark.createDataFrame(spark.sparkContext.parallelize(data), schema)
df.show
+--------+
|      c1|
+--------+
|  [, a2]|
|[b1, b2]|
|    null|
+--------+
```
In Spark 2.4.4,
```
df.na.drop("any", Seq("c1.c1-1")).show
+--------+
|      c1|
+--------+
|[b1, b2]|
+--------+
```
In Spark 2.4.5 or Spark 3.0.0-preview2, if nested columns are specified, they are ignored.
```
df.na.drop("any", Seq("c1.c1-1")).show
+--------+
|      c1|
+--------+
|  [, a2]|
|[b1, b2]|
|    null|
+--------+
```
### Why are the changes needed?

This seems like a regression.

### Does this PR introduce any user-facing change?

Now, the nested column can be specified:
```
df.na.drop("any", Seq("c1.c1-1")).show
+--------+
|      c1|
+--------+
|[b1, b2]|
+--------+
```

Also, if `*` is specified as a column, it will throw an `AnalysisException` that `*` cannot be resolved, which was the behavior in 2.4.4. Currently, in master, it has no effect.

### How was this patch tested?

Updated existing tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala [hunk 1]
```diff
@@ -275,33 +285,35 @@ class DataFrameNaFunctionsSuite extends QueryTest with SharedSparkSession {
...
-    val schema = new StructType()
-      .add("c1", new StructType()
-        .add("c1-1", StringType)
-        .add("c1-2", StringType))
+  test("drop with col(*)") {
+    val df = createDF()
+    val exception = intercept[AnalysisException] {
+      df.na.drop("any", Seq("*"))
+    }
+    assert(exception.getMessage.contains("Cannot resolve column name \"*\""))
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala:299]: Note that this was the behavior in Spark 2.4.4. We can handle this more gracefully (e.g., use `outputAttributes`) if we need to.

On a side note, for `fill`, `*` is ignored in Spark 2.4.4.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala [hunk 2]
```diff
@@ -275,33 +285,35 @@ class DataFrameNaFunctionsSuite extends QueryTest with SharedSparkSession {
...
-      Row(Row(null, "a2")),
-      Row(Row("b1", "b2")),
-      Row(null))
+  test("fill with nested columns") {
+    val df = createDFWithNestedColumns
 
-    val df = spark.createDataFrame(
-      spark.sparkContext.parallelize(data), schema)
+    // Nested columns are ignored for fill().
+    checkAnswer(df.na.fill("a1", Seq("c1.c1-1")), df)
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala:306]: Note that nested columns are ignored for `fill` in Spark 2.4.4.

**Comment by imback82** [comment]
@cloud-fan Please let me know if this PR (going back to 2.4.4 behavior) makes sense. Thanks!

**Comment by TJX2014** [comment]
nice,duplicate columns is same as struct alias which not works at toAttributes method in DataFrameNaFunctions.

**Review by cloud-fan — APPROVED** [review]
good catch!

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala [hunk 3]
```diff
@@ -275,33 +285,35 @@ class DataFrameNaFunctionsSuite extends QueryTest with SharedSparkSession {
...
+  }
 
-    checkAnswer(df.select("c1.c1-1"),
-      Row(null) :: Row("b1") :: Row(null) :: Nil)
+  test("drop with nested columns") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala:309]: nit: This looks like a bug, so could you add the prefix: `SPARK-31256`.

**Review by maropu — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala (hunk 3 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala:309]: nvm, a bit late...

**Comment by cloud-fan** [comment]
merging to master/3.0/2.4

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameNaFunctionsSuite.scala:309]: 😂

**Comment by dongjoon-hyun** [comment]
So, SPARK-31256 made a regression at 2.4.5 and this recovers it?

**Comment by cloud-fan** [comment]
@dongjoon-hyun yes

**Comment by dongjoon-hyun** [comment]
Thank you for confirmation~

**Pull request closed without merging**